_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+same platform for both application and infrastructure CI/CD workflows. In addition, we recommend selecting a Saas
+based CI/CD platform to avoid the overhead of managing additional infrastructure for the VCS integration and workflow
+execution.
+
+However, instead of relying on the CI/CD platform to run infrastructure deployments (e.g calling `terraform apply`), use
+a specialized self-hosted deployment platform for running your infrastructure deployments. This can be Terraform
+Enterprise if you are only using `terraform`, or Gruntwork's ECS Deploy Runner stack for a more general purpose
+platform. In order to support arbitrary infrastructure deployments, it will be inevitable that the system running your
+deployments will have de factor admin status to the environments that you are deploying into. As such, it is important
+that these systems are locked down. You do not want any third party publicly facing system to hold powerful credentials
+to your cloud environments.
+
+This design implements separates out the concerns so that we take full advantage of the strengths of each platform,
+while covering the weaknesses: relying on the CI/CD platforms to manage the workflow/pipline, but have it trigger
+infrastructure deployments on self-hosted systems that are more locked down. In this model, the CI/CD platforms only
+need enough permissions to trigger a deployment, but not necessarily all the permissions required for running a
+deployment. Instead, you can have a system that is closed off in your isolated network hold the credentials
+necessary to actually run the deployment.
+
+
+[[use_serverless_platforms_for_infrastructure_deployments]]
+=== Use serverless platforms for infrastructure deployments
+
+For running infrastructure deployments, choose a system that makes use of serverless platforms. For example, Gruntwork's
+ECS Deploy Runner utilizes ECS Fargate to run the infrastructure code, with AWS Lambda as a frontend to provide a
+callable function (driven by the AWS SDK and authorization managed by IAM). By utilizing Fargate and AWS Lambda, we have
+a purely serverless architecture where no infrastructure needs to be managed, and IAM credentials are short lived. This
+not only helps reduce the overhead of running the system, but it also helps reduce the attack surface (e.g you don't
+have to worry about a permanent, externally available server that needs to be constantly patched).
+
+
+[[use_a_vpc_to_lock_down_infrastructure_deployer]]
+=== Use a VPC to lock down infrastructure deployer
+
+Run your infrastructure deployment workloads in a https://aws.amazon.com/vpc/[Virtual Private Cloud (VPC)] to isolate
+the workloads in a restricted network topology (see link:/guides/networking/how-to-deploy-production-grade-vpc-aws[How
+to deploy a production-grade VPC on AWS] for more information on VPCs). Configure to run all workloads in private
+subnets that are not publicly accessible.
```

Comment:
Block all inbound Internet access and consider blocking all outbound access except for the minimum required (e.g,. allow access to AWS APIs).

Reply:
Updated: 427a56bc1a9eb400331fdb34cf82e44a9d6aabcb